### PR TITLE
Some conversion work

### PR DESCRIPTION
**Pull request by KrissN**
Hi,

I've started some e4x->xml2jxon conversion work. This is the code I've converted so far.

You're free to merge it if you find it useful.

### CODE REVIEW

**Comment by 1stsetup** [comment]
Op 04-09-12 15:26, Krzysztof Nowicki schreef:

> Hi,
> 
> I've started some e4x->xml2jxon conversion work. This is the code I've
> converted so far.
> 
> You're free to merge it if you find it useful.
> 
> ---
> 
> ```
>     You can merge this Pull Request by running:
> ```
> 
>   git pull https://github.com/KrissN/exchangecalendar e4xReplacementKrissN
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/1stsetup/exchangecalendar/pull/2
> 
> ```
>     Commit Summary
> ```
> - Converted erSyncFolderItems to new xml2jxon object
> - Converted erFindMasterOccurrences to new xml2jxon object
> - Converted erGetItems to new xml2jxon object
> - Converted calExchangeCalendar::convertExchangeToCal (+ friends) to ne…
>   
>   ```
>   File Changes
>   ```
> - _M_ components/calExchangeCalendar.js (179)
> - _M_ components/erFindMasterOccurrences.js (286)
> - _M_ components/erGetItems.js (276)
> - _M_ components/erSyncFolderItems.js (73)
>   
>   ```
>   Patch Links
>   ```
> - https://github.com/1stsetup/exchangecalendar/pull/2.patch
> - https://github.com/1stsetup/exchangecalendar/pull/2.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/1stsetup/exchangecalendar/pull/2.
> 
> Thank you. I will have a look at it tonight.

Michel.

**Comment by 1stsetup** [comment]
Krzysztof,

I need some more time to evaluate. But it really looks all right. This was next on my list to do. No I have more time for the update of calendar items from TB to Exchange.

Would you mind to add your name as contributor to the head of the file?
Regards,

Michel.

**Comment by KrissN** [comment]
Thanks.

I've added my name to the file headers.

The current code starts to work now (I can see some calendar items). My next target is erFindOccurencies.

Eventually I'll also try to add a Polish translation, since there isn't any ATM.

Regards
Chris

**Comment by 1stsetup** [comment]
Chris,

I tested it also in TB15 and it works. It only works on reading items as soon as you change an item it triggers the E4X errors but that is not strange because that still needs to be done. tonight I have time again to start on that and check the last of your code.

File: components/calExchangeCalendar.js [hunk 1]
```diff
@@ -6727,15 +6735,15 @@ this.logInfo("getTaskItemsOK 4");
...
+			//this.logInfo(" == item.title:"+item.title+", calendarItemType:"+aCalendarItem.getTagValue("t:CalendarItemType"));
+			switch (aCalendarItem.getTagValue("t:CalendarItemType")) {
 				case "Exception" :
 					this.logInfo("@1:"+item.startDate.toString()+":IsException");
 					item.setProperty("X-RecurringType", "RE");
 					this.setAlarm(item, aCalendarItem);  
 					// Try to find master. If found add Exception and link recurrenceinfo.
-					item.recurrenceId = this.tryToSetDateValue(aCalendarItem.nsTypes::RecurrenceId, item.startDate);
-					var master = this.recurringMasterCache[aCalendarItem.nsTypes::UID.toString()];
+					item.recurrenceId = this.tryToSetDateValue(aCalendarItem.getTagValue("t::RecurrenceId"), item.startDate);
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: This needs to be changed into "t:RecurrenceId"

File: components/calExchangeCalendar.js [hunk 2]
```diff
@@ -6597,18 +6602,19 @@ this.logInfo("getTaskItemsOK 4");
 */
 
 		var cats = [];
-		for each (var cat in aCalendarItem.nsTypes::Categories.nsTypes::String) {
-			cats.push(cat.toString());
+//		for each (var cat in aCalendarItem.nsTypes::Categories.nsTypes::String) {
+		for each (var cat in aCalendarItem.XPath("/t:Categories/t:String")) {
+			cats.push(cat.value());
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: This one needs to be changed into: cats.push(cat.value);
Your version generates a javascript error and causes the spinning arrow (job count) to stay active.

File: components/calExchangeCalendar.js [hunk 3]
```diff
@@ -6244,11 +6247,11 @@ this.logInfo("getTaskItemsOK 4");
 				break;
 		}
 
-		if (aExchangeItem.nsTypes::IsCancelled.toString() == "true") {
+		if (aExchangeItem.getTagValue("IsCancelled") == "true") {
 			this.setStatus(aItem, "Decline");
 		}
 		else {
-			this.setStatus(aItem, aExchangeItem.nsTypes::MyResponseType.toString());
+			this.setStatus(aItem, aExchangeItem.getTagValue("MyResponseType"));
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Please add the namespace "t:MyResponseType"

File: components/calExchangeCalendar.js [hunk 4]
```diff
@@ -6244,11 +6247,11 @@ this.logInfo("getTaskItemsOK 4");
 				break;
 		}
 
-		if (aExchangeItem.nsTypes::IsCancelled.toString() == "true") {
+		if (aExchangeItem.getTagValue("IsCancelled") == "true") {
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Please add the namespace "t:IsCancelled"

File: components/calExchangeCalendar.js [hunk 5]
```diff
@@ -6229,7 +6232,7 @@ this.logInfo("getTaskItemsOK 4");
 				aItem.privacy = "PUBLIC";
 		}
 
-		switch (aExchangeItem.nsTypes::LegacyFreeBusyStatus.toString()) {
+		switch (aExchangeItem.getTagValue("LegacyFreeBusyStatus")) {
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Please add the namespace "t:LegacyFreeBusyStatus"

File: components/calExchangeCalendar.js [hunk 6]
```diff
@@ -6212,7 +6215,7 @@ this.logInfo("getTaskItemsOK 4");
 				break;
 		}
 
-		switch (aExchangeItem.nsTypes::Sensitivity.toString()) {
+		switch (aExchangeItem.getTagValue("Sensitivity")) {
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Please add the namespace "t:Sensitivity"

File: components/calExchangeCalendar.js [hunk 7]
```diff
@@ -6200,7 +6203,7 @@ this.logInfo("getTaskItemsOK 4");
 
 	setCommonValues: function _setCommonValues(aItem, aExchangeItem)
 	{
-		switch(aExchangeItem.nsTypes::Importance.toString()) {
+		switch(aExchangeItem.getTagValue("Importance")) {
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Please add the namespace "t:Importance"

File: components/calExchangeCalendar.js [hunk 8]
```diff
@@ -5987,25 +5989,25 @@ this.logInfo("getTaskItemsOK 4");
 
 	createAttendee: function _createAttendee(aElement, aType, aMyResponseType) 
 	{
-		let mbox = aElement.nsTypes::Mailbox;
+		let mbox = aElement["t:Mailbox"];
```
Comment (1stsetup) [inline components/calExchangeCalendar.js]: Maybe better we change this to
let mbox = aElement.getTag("t:Mailbox");

File: components/erFindMasterOccurrences.js [hunk 9]
```diff
@@ -88,128 +90,159 @@ erFindMasterOccurrencesRequest.prototype = {
...
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:Body");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ResponseObjects");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:Sensitivity");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderDueBy");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderIsSet");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderMinutesBeforeStart");
+
+		var extFieldURI;
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erFindMasterOccurrences.js]: ident is differen than the rest.

File: components/erFindMasterOccurrences.js [hunk 10]
```diff
@@ -88,128 +90,159 @@ erFindMasterOccurrencesRequest.prototype = {
...
+		extFieldURI.setAttribute("PropertyId", MAPI_PidLidReminderSignalTime);
+		extFieldURI.setAttribute("PropertyType", "SystemTime");
+
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erFindMasterOccurrences.js]: ident is differen than the rest.

File: components/erFindMasterOccurrences.js [hunk 11]
```diff
@@ -88,128 +90,159 @@ erFindMasterOccurrencesRequest.prototype = {
...
+		extFieldURI.setAttribute("PropertyId", MAPI_PidLidReminderSet);
+		extFieldURI.setAttribute("PropertyType", "Boolean");
+
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erFindMasterOccurrences.js]: ident is differen than the rest.

File: components/erFindMasterOccurrences.js [hunk 12]
```diff
@@ -88,128 +90,159 @@ erFindMasterOccurrencesRequest.prototype = {
...
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:Mileage");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:Owner");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:PercentComplete");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:Recurrence");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:StartDate");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:Status");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:StatusDescription");
+			additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "task:TotalWork");
+
+	    	extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
```
Comment (1stsetup) [inline components/erFindMasterOccurrences.js]: ident is differen than the rest for next few lines.

File: components/erGetItems.js [hunk 13]
```diff
@@ -83,128 +85,154 @@ erGetItemsRequest.prototype = {
...
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:Body");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ResponseObjects");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:Sensitivity");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderDueBy");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderIsSet");
+		additionalProperties.addChildTag("FieldURI", "nsTypes", null).setAttribute("FieldURI", "item:ReminderMinutesBeforeStart");
+			    
+		var extFieldURI;
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erGetItems.js]: ident is differen than the rest.

File: components/erGetItems.js [hunk 14]
```diff
@@ -83,128 +85,154 @@ erGetItemsRequest.prototype = {
...
+		extFieldURI.setAttribute("PropertyId", MAPI_PidLidReminderSignalTime);
+		extFieldURI.setAttribute("PropertyType", "SystemTime");
 
-		req.nsMessages::ItemShape.nsTypes::AdditionalProperties.content += <>
-		    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Common" PropertyId={MAPI_PidLidReminderSignalTime} PropertyType="SystemTime" xmlns:nsTypes={nsTypes}/>
-		    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Common" PropertyId={MAPI_PidLidReminderSet} PropertyType="Boolean" xmlns:nsTypes={nsTypes}/>
-		    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Common" PropertyId={MAPI_PidLidReminderDelta} PropertyType="Integer" xmlns:nsTypes={nsTypes}/>
-		    </>;
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erGetItems.js]: ident is differen than the rest.

File: components/erGetItems.js [hunk 15]
```diff
@@ -83,128 +85,154 @@ erGetItemsRequest.prototype = {
...
+		extFieldURI.setAttribute("PropertyId", MAPI_PidLidReminderSet);
+		extFieldURI.setAttribute("PropertyType", "Boolean");
 
+		extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
+	    extFieldURI.setAttribute("DistinguishedPropertySetId", "Common");
```
Comment (1stsetup) [inline components/erGetItems.js]: ident is differen than the rest.

File: components/erGetItems.js [hunk 16]
```diff
@@ -83,128 +85,154 @@ erGetItemsRequest.prototype = {
...
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskAccepted} PropertyType="Boolean" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskLastUpdate} PropertyType="SystemTime" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskAcceptanceState} PropertyType="Integer" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskMode} PropertyType="Integer" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskGlobalId} PropertyType="Binary" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskHistory} PropertyType="Integer" xmlns:nsTypes={nsTypes}/>
-			    	<nsTypes:ExtendedFieldURI DistinguishedPropertySetId="Task" PropertyId={MAPI_PidLidTaskOwnership} PropertyType="Integer" xmlns:nsTypes={nsTypes}/>
-			    </>;
+	    	extFieldURI = additionalProperties.addChildTag("ExtendedFieldURI", "nsTypes", null);
```
Comment (1stsetup) [inline components/erGetItems.js]: ident is differen than the rest for the following lines

File: components/erGetItems.js [hunk 17]
```diff
@@ -241,34 +269,46 @@ erGetItemsRequest.prototype = {
...
+		for each (var e in rm) {
+			var item = e.XPath("/m:Items/t:CalendarItem");
+			if (item.length > 0)
+			{
+				exchWebService.commonFunctions.LOG("erGetTaskItemsRequest.item: "+item[0]+"\n");
+				items.push(item[0]);
+			}
 		}
-	
+
```
Comment (1stsetup) [inline components/erGetItems.js]: White space. Please remove.

**Comment by 1stsetup** [comment]
Chris. Please see my comments. I hope you have time to add requested changes. If so please update the pull request and I will incorporate it into the e4xReplacement branche.

**Comment by KrissN** [comment]
I've fixed the issues you reported.

**Pull request merged**